packages/lime-proto-babeld/files/usr/lib/lua/lime/proto/babeld.lua

```diff
@@ -103,21 +114,70 @@ function babeld.setup_interface(ifname, args)
 	local vlanProto = args[3] or "8021ad"
 	local nameSuffix = args[4] or "_babeld"
 
+
+	--! If Babeld is without VLAN (vlanId is 0) it should run directly on plain
+	--! ethernet interfaces, but the ones which are inside of the LAN bridge
+	--! (e.g. eth0 or eth0.1) cannot have an IPv6 Link-Local and Babeld needs it.
+	--! So Babeld has to run on the bridge interface br-lan
+	local isIntoLAN = false
+	local addIPtoIf = true
+	for _,v in pairs(args["deviceProtos"]) do
+		if v == "lan" then
+			isIntoLAN = true
+		--! would be weird to add a static IP to the WAN interface
+		elseif v == "wan" then
+			addIPtoIf = false
+		end
+	end
+
+	if ifname:match("^wlan") then
+		--! currently (2019-10-12) mode-ap and mode-apname have an hardcoded
+		--! "option network lan" so they are always in the br-lan bridge
+		if ifname:match("^wlan.*ap$") or ifname:match("^wlan.*apname$") then
+			isIntoLAN = true
+
+		--! all the WLAN interfaces are ignored by proto-lan
+		--! so they are not in the bridge even if proto-lan is present
+		--! (except mode-ap and mode-apname as mentioned above)
+		else
+			isIntoLAN = false
+		end
+	end
+
+	if tonumber(vlanId) == 0 and isIntoLAN then
+		utils.log("Rather than "..ifname..
+		  ", adding br-lan into Babeld interfaces")
+		ifname = "br-lan"
+		--! br-lan has already an IPv4, no need to add it
+		addIPtoIf = false
+	end
+
 	local owrtInterfaceName, linuxVlanIfName, owrtDeviceName =
 	  network.createVlanIface(ifname, vlanId, nameSuffix, vlanProto)
 
-	local ipv4, _ = network.primary_address()
-
 	local uci = config.get_uci_cursor()
 
-	if(vlanId ~= 0 and ifname:match("^eth")) then
+	if tonumber(vlanId) ~= 0 and ifname:match("^eth") then
 		uci:set("network", owrtDeviceName, "mtu", tostring(network.MTU_ETH_WITH_VLAN))
 	end
 
-	uci:set("network", owrtInterfaceName, "proto", "static")
-	uci:set("network", owrtInterfaceName, "ipaddr", ipv4:host():string())
-	uci:set("network", owrtInterfaceName, "netmask", "255.255.255.255")
-	uci:save("network")
+	if addIPtoIf then
+		local ipv4, _ = network.primary_address()
+		--! the "else" way should always work but it fails in a weird way
+		--! with some wireless interfaces without VLAN
+		--! (e.g. works with wlan0-mesh and fails with wlan1-mesh)
+		--! so for these cases, the first way is used
+		--! (which indeed fails for most of the other cases)
+		if ifname:match("^wlan") and tonumber(vlanId) == 0 then
+			uci:set("network", owrtInterfaceName, "ifname", "@"..owrtDeviceName)
+		else
+			uci:set("network", owrtInterfaceName, "ifname", linuxVlanIfName)
+		end
```

Comment:
I did test it on 5 different routers (listed in the first comment in the PR), but more testing would also be nice to have :)
Can you point me to the other places in the code where such a workaround is used (so that I can compare solutions)?

Reply:
> > > `babeld_over_batman` option should not be removed, is what permit us to eventually setup and test babel only on frontier nodes, as per the original libremesh scalability proposal
> > 
> > 
> > @G10h4ck ah... Rly? This scenario is the number 1 feature in LibreMesh and I did not understand how it should work until now. More documentation is needed (this is not explained neither in the _How it works_). I will add that option back as soon as possible.
> 
> To implement libremesh full proposal is not a simple task, so we started from the easier part, which is solvable more or less just by generating a static configuration when lime-config is run. The other parts of the proposal (for example automatically turn on L3 routing protocol only on frontier nodes) needs more thins going on at run-time, and we are implementing those things bit by bit (shared-state is one of those things). In the meanwhile even if we don't have all those pieces in place is important to keep the pieces we already have compatible with that scenario, so when we get there it is not too overwhelming to make it work.

I just added back the babeld_over_batman option, I just took the code for it from #600 so it should work (not tested now, but I tested some times ago in the previous PR).

Reply:
@spiccinini could you test this on LibreRouter? Two scenarios should be tested: with VLAN and without VLAN.
To verify that it works: issue `echo dump | nc ::1 30003` and check if all the included interfaces (apart from the WAN one) have an IPv4 in both VLAN and noVLAN configurations. In the noVLAN configuration the br-lan interface has to be present in the list of the Babeld interfaces.

Reply:
I've tested with no luck:

```
# with VLAN 17
root@nodo1:~# echo dump | nc ::1 30003
BABEL 1.0
version babeld-1.9.1
host nodo1
my-id aa:40:41:ff:fe:1c:85:6e
ok
add interface eth0-1_17 up true ipv6 fe80::aa40:41ff:fe1c:856e
add interface eth0.1 up true ipv6 fe80::aa40:41ff:fe1c:856e ipv4 10.0.0.228
add interface eth0 up true ipv6 fe80::aa40:41ff:fe1c:856e
add interface eth0_17 up true ipv6 fe80::aa40:41ff:fe1c:856e ipv4 10.85.133.110
add interface eth1-2_17 up true ipv6 fe80::aa40:41ff:fe1c:856f ipv4 10.85.133.110
add interface eth1.2 up false
add interface eth1 up true ipv6 fe80::aa40:41ff:fe1c:856f
add interface eth1_17 up true ipv6 fe80::aa40:41ff:fe1c:856f ipv4 10.85.133.110
add interface wlan0-mesh up true ipv6 fe80::aa40:41ff:fe1c:856c
add interface wlan0-mesh_17 up false
add interface wlan1-mesh up true ipv6 fe80::aa40:41ff:fe1c:83f3
add interface wlan1-mesh_17 up false
add interface wlan2-mesh up true ipv6 fe80::aa40:41ff:fe1c:83f9
add interface wlan2-mesh_17 up false
add xroute 2a00:1508:a55:cd00::/64-::/0 prefix 2a00:1508:a55:cd00::/64 from ::/0 metric 0
add xroute 0.0.0.0/0-0.0.0.0/0 prefix 0.0.0.0/0 from 0.0.0.0/0 metric 0
add xroute 10.85.0.0/16-0.0.0.0/0 prefix 10.85.0.0/16 from 0.0.0.0/0 metric 0
add xroute 10.0.0.0/24-0.0.0.0/0 prefix 10.0.0.0/24 from 0.0.0.0/0 metric 0
ok

With VLAN=0
```
NO VLAN config
root@nodo1:~# echo dump | nc ::1 30003
BABEL 1.0
version babeld-1.9.1
host nodo1
my-id aa:40:41:ff:fe:1c:85:6e
ok
add interface eth0.1 up true ipv6 fe80::aa40:41ff:fe1c:856e ipv4 10.0.0.228
add interface eth0 up true ipv6 fe80::aa40:41ff:fe1c:856e ipv4 10.85.133.110
add interface eth1.2 up false
add interface eth1 up true ipv6 fe80::aa40:41ff:fe1c:856f ipv4 10.85.133.110
add interface wlan0-mesh up true ipv6 fe80::aa40:41ff:fe1c:856c
add interface wlan1-mesh up true ipv6 fe80::aa40:41ff:fe1c:83f3
add interface wlan2-mesh up true ipv6 fe80::aa40:41ff:fe1c:83f9
add neighbour 77fddde0 address fe80::aa40:41ff:fe1c:83f3 if wlan2-mesh reach 000e ureach 0000 rxcost 65535 txcost 65535 rtt 0.000 rttcost 0 cost 65535
add xroute 0.0.0.0/0-0.0.0.0/0 prefix 0.0.0.0/0 from 0.0.0.0/0 metric 0
add xroute 10.0.0.0/24-0.0.0.0/0 prefix 10.0.0.0/24 from 0.0.0.0/0 metric 0
ok
```
some messages:
```
Mon Nov 11 22:38:17 2019 daemon.err babeld[1690]: Interface eth1.2 has no link-local address.
Mon Nov 11 22:38:17 2019 daemon.err babeld[1690]: setsockopt(IPV6_LEAVE_GROUP): Address not available

35: eth1.2@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP group default qlen 1000
    link/ether a8:40:41:1c:85:6f brd ff:ff:ff:ff:ff:ff

```

When using vlan=0 ipv4 stops working (no DHCP, no DNS)

Reply:
@spiccinini can you confirm that you used the code from this PR?
Looks like the fix did not work at all (or it was not present).

Reply:
Yes I used the code from this PR (rebased to master). Maybe I did a mistake. I deleted both /etc/config wireless and network. Regenerated them with generate_config, then lime-config and lime-apply,

Reply:
I am not familiar with generate_config and I never trusted lime-apply...
Maybe #75 is related?
@spiccinini could you try without deleting the config files, and just running lime-config and rebooting? Just for being sure...

Reply:
If I don't delete the config the vlan 17 still exists after lime-config

Reply:
Wow... Even after the reboot? So something is failing baaaad.
Can you post the full /etc/config/network with in the two cases of VLAN set in /etc/config/lime
 and VLAN set to zero in /etc/config/lime?

Reply:
I did more testing and for me it works. Can you confirm that it fails with LibreRouter or with your compilation procedure?
Regarding the tests, by myself I am not able to write them, can you do it or guide me?